### PR TITLE
updated deps to get gtkplus to build

### PR DESCRIPTION
**Pull request by lee218llnl**

### CODE REVIEW

**Comment by aweits** [comment]
Hm, this pulls rust in, which is currently not building for me... I take it this makes macos work?

**Comment by adamjstewart** [comment]
Testing this on macOS now. For rust, I would recommend using a binary installer and registering it as an external package if you only need it as a dependency of a dependency.

**Comment by adamjstewart** [comment]
I actually get an error while trying to build `gtk-doc`:
```
1 error found in build log:
     135    checking for dblatex... no
     136    checking for fop... no
     137    configure: WARNING: neither dblatex nor fop found, so no pdf output from xml
     138    checking for XML catalog... not found
     139    checking for xmlcatalog... /Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatal
            og
     140    checking for DocBook XML DTD V4.3 in XML catalog... not found
  >> 141    configure: error: could not find DocBook XML DTD V4.3 in XML catalog
```

**Comment by lee218llnl** [comment]
> I actually get an error while trying to build `gtk-doc`:
> 
> ```
> 1 error found in build log:
>      135    checking for dblatex... no
>      136    checking for fop... no
>      137    configure: WARNING: neither dblatex nor fop found, so no pdf output from xml
>      138    checking for XML catalog... not found
>      139    checking for xmlcatalog... /Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatal
>             og
>      140    checking for DocBook XML DTD V4.3 in XML catalog... not found
>   >> 141    configure: error: could not find DocBook XML DTD V4.3 in XML catalog
> ```

On our systems it was finding the OS-installed dependences. I just pushed a commit that will have Spack pull its own builds. @adamjstewart can you please pull with the latest commit and try again?

**Comment by adamjstewart** [comment]
Even with the latest commit I still see the same error.

**Comment by lee218llnl** [comment]
> Even with the latest commit I still see the same error.

I just pushed another commit. It adds the specific version of docbook-xsl that gtk-doc needs and also patches gtk-docs configure macro. The way the configure script worked was that it was being passed a single XML catalog file to xmlcatalog, but with the spack installation we have separate ones for docbook-xml and docbook-xsl. The xmlcatalog file would normally honor the XML_CATALOG_FILES environment variable if not for the fact that gtk-doc's configure was specifying a singular one. Anyway, please try again. This is what I see in my successful config.log:

configure:12630: result: /etc/xml/catalog
configure:12641: checking for xmlcatalog
configure:12659: found /usr/WS2/lee218/delete/spack-gtk/opt/spack/linux-rhel7-broadwell/gcc-8.1.0/libxml2-2.9.10-wtwzmbpxll62byvru27uaes56vcc5v4v/bin/xmlcatalog
configure:12672: result: /usr/WS2/lee218/delete/spack-gtk/opt/spack/linux-rhel7-broadwell/gcc-8.1.0/libxml2-2.9.10-wtwzmbpxll62byvru27uaes56vcc5v4v/bin/xmlcatalog
configure:12691: checking for DocBook XML DTD V4.3 in XML catalog
configure:12694: $XMLCATALOG --noout "" "-//OASIS//DTD DocBook XML V4.3//EN" >&2
/usr/WS2/lee218/delete/spack-gtk/opt/spack/linux-rhel7-broadwell/gcc-8.1.0/docbook-xml-4.3-darcytdof354dlddjppxmju6ekg4ebuo/docbookx.dtd
configure:12697: $? = 0
configure:12699: result: found
configure:12709: checking for DocBook XSL Stylesheets in XML catalog
configure:12712: $XMLCATALOG --noout "" "http://docbook.sourceforge.net/release/xsl/current/html/chunk.xsl" >&2
/usr/WS2/lee218/delete/spack-gtk/opt/spack/linux-rhel7-broadwell/gcc-8.1.0/docbook-xsl-1.78.1-n7zahsv5l2hf2igpgamtnbjxhqixdedc/html/chunk.xsl

**Comment by adamjstewart** [comment]
Hmm, still getting the same error even with the patch. Contents of `config.log`:
```
configure:12603: checking for XML catalog
configure:12633: result: not found
configure:12641: checking for xmlcatalog
configure:12659: found /Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatalog
configure:12672: result: /Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatalog
configure:12691: checking for DocBook XML DTD V4.3 in XML catalog
configure:12703: result: not found
configure:12705: error: could not find DocBook XML DTD V4.3 in XML catalog
```

**Comment by lee218llnl** [comment]
> Hmm, still getting the same error even with the patch. Contents of `config.log`:
> 
> ```
> configure:12603: checking for XML catalog
> configure:12633: result: not found
> configure:12641: checking for xmlcatalog
> configure:12659: found /Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatalog
> configure:12672: result: /Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatalog
> configure:12691: checking for DocBook XML DTD V4.3 in XML catalog
> configure:12703: result: not found
> configure:12705: error: could not find DocBook XML DTD V4.3 in XML catalog
> ```

Did the build.patch get applied? Can you attach the full config.log?

**Comment by adamjstewart** [comment]
The patch did get applied:
```
==> Applied patch /Users/Adam/spack/var/spack/repos/builtin/packages/gtk-doc/build.patch
```
Full build log and config log:

* [spack-build-out.txt](https://github.com/spack/spack/files/6152602/spack-build-out.txt)
* [config.log](https://github.com/spack/spack/files/6152604/config.log)

**Comment by lee218llnl** [comment]
> The patch did get applied:
> 
> ```
> ==> Applied patch /Users/Adam/spack/var/spack/repos/builtin/packages/gtk-doc/build.patch
> ```
> 
> Full build log and config log:
> 
>     * [spack-build-out.txt](https://github.com/spack/spack/files/6152602/spack-build-out.txt)
> 
>     * [config.log](https://github.com/spack/spack/files/6152604/config.log)

Nothing is standing out in those 2 files. Can you please also attach your spack-build-env.txt and configure? Thanks for your help and sorry for all the iterations here. I'm on a PC and don't have access to a Mac, so I cannot reproduce what you are seeing.

**Comment by adamjstewart** [comment]
* [spack-build-env.txt](https://github.com/spack/spack/files/6153587/spack-build-env.txt)
* [configure](https://github.com/spack/spack/files/6153589/configure.txt)

**Comment by lee218llnl** [comment]
@adamjstewart Can you run an interactive test for me? If so, please set:

XML_CATALOG_FILES='/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/docbook-xsl-1.78.1-hzs6ux2yxnyhnfjnzuvd6vfxdsl376b7/catalog.xml /Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/docbook-xml-4.3-upmmx3mdyopzjkzbuav6en4exodfs2la/catalog.xml'

And then run:

/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatalog --noout "" "-//OASIS//DTD DocBook XML V4.3//EN"

and 

/Users/Adam/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/libxml2-2.9.10-dqzlvh6klvk75t2gw7qnt64s3bspzozx/bin/xmlcatalog --noout "" "http://docbook.sourceforge.net/release/xsl/current/html/chunk.xsl"

And can you also please attach to two files listed in XML_CATALOGS_FILES?

**Comment by lee218llnl** [comment]
Ah, I see what the issue is now. On our system, it finds /etc/xml/catalog by default, but since your system does not have it, you see:

> configure:12603: checking for XML catalog
> configure:12633: result: not found

In the absence of finding such a file, the configure script doesn't run the xmlcatalog test to look for the 2 catalogs. I will see if I can further patch the build system.

**Comment by lee218llnl** [comment]
@adamjstewart sorry, I need to have you run another test on the latest commit. I modified the patch, so you may need to clean your ~/.spack/cache/patches directory.

**Comment by adamjstewart** [comment]
Thanks @lee218llnl, the patch now allows me to build `gtk-doc`! I had to add another commit to fix a `libcroco` build error on macOS (hopefully this change is also fine on Linux). I also had to add a new version of `librsvg` to get it to build on macOS. I'm currently stuck at the same `gtkplus` build error I was seeing before. I attached the logs if you want to take a look.

* [spack-build-out.txt](https://github.com/spack/spack/files/6158286/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/6158287/spack-build-env.txt)

**Comment by lee218llnl** [comment]
I'm not going to be much help for the Mac OS gtkplus build. Would it be OK to merge this PR and open a separate PR/issue for the Mac OS build? Or does this PR somehow inherently break gtkplus on Mac OS? Can you build gtkplus from the develop branch?

**Comment by adamjstewart** [comment]
gtkplus was already broken on develop, so I'm fine with merging this as is and opening a separate issue.

File: var/spack/repos/builtin/packages/docbook-xml/package.py [hunk 1]
```diff
@@ -15,6 +15,8 @@ class DocbookXml(Package):
     list_url = "https://www.oasis-open.org/docbook/xml/"
     list_depth = 1
 
+    version('4.3', sha256='23068a94ea6fd484b004c5a73ec36a66aa47ea8f0d6b62cc1695931f5c143464')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/docbook-xml/package.py]: Can you sort these versions from newest to oldest?

File: var/spack/repos/builtin/packages/docbook-xsl/package.py [hunk 2]
```diff
@@ -13,6 +13,7 @@ class DocbookXsl(Package):
     homepage = "https://github.com/docbook/xslt10-stylesheets"
     url      = "https://github.com/docbook/xslt10-stylesheets/releases/download/release%2F1.79.2/docbook-xsl-1.79.2.tar.bz2"
 
+    version('1.78.1', sha256='c98f7296ab5c8ccd2e0bc07634976a37f50847df2d8a59bdb1e157664700b467', url='https://sourceforge.net/projects/docbook/files/docbook-xsl/1.78.1/docbook-xsl-1.78.1.tar.bz2')
     version('1.79.2', sha256='316524ea444e53208a2fb90eeb676af755da96e1417835ba5f5eb719c81fa371')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/docbook-xsl/package.py:16]: Same here

**Review by adamjstewart — APPROVED** [review]

**Comment by lee218llnl** [comment]
@adamjstewart I'm not sure what is going wrong with the Python 3.8 build. It looks like it's just failing to clean up. @becker33 @tldahlgren any ideas?

**Pull request merged**